Asset Swaps for ZSAs

vivek-arte opened the pull request:
We propose **ZIP 228: Zcash Shielded Asset Swaps** in this PR.

This ZIP is to be reviewed and implemented for the Asset Swaps protocol over Orchard-ZSA (which is specified in ZIP 226 and ZIP 227).

The rendered HTML version of ZIP 228 can be viewed [here](https://deploy-preview-50--zcash-zips-qedit.netlify.app/zip-0228).

The target issue for Asset Swaps for ZSAs (ZIP 228) is https://github.com/zcash/zips/issues/776

vivek-arte committed:
creation of Asset Swaps ZIP based on Google Doc Draft

daira commented on zip-0228.rst:
```suggestion
By including :math:`\mathsf{bsk}` values in each Swap Order, and building the transaction :math:`\mathsf{bsk}` from these values as described in the `Binding Signature Changes`_ section, we prevent this malicious behavior.
```

vivek-arte replied:
Oops. Fixed this along with other smaller updates in [this commit](https://github.com/zcash/zips/pull/780/commits/69ff7a8dd7537dadc1c69031b1d44a66969b6962).

vivek-arte committed:
Fixing typo, clarifying information about Partial Fills

vivek-arte committed:
correcting bytes[] to byte[]

vivek-arte committed:
making changes based on review comments

vivek-arte committed:
updates to ZIP 228 files and readme post restructuring changes

vivek-arte committed:
updating path to images post restructuring so that images render correctly

vivek-arte committed:
Updates to Asset Swaps ZIP based on review comments (#62)

This makes updates based on the comments made on #50.

vivek-arte committed:
updating references to protocol specification

ConstanceBeguier committed:
[Swaps] Update swap bundle figure (#74)

In swap bundle figure,
- replace spend and output notes by actions
- fix bundle value balance

ConstanceBeguier committed:
[ZIP228] Simplify swap bundle figure

- fix global bundle value balance by adding fees
- remove action fields (cv_net, nf, rk, cm_x, epk, ...)  for each action
- remove global transaction fields (header, nVersionGroupId,
nConsensusBranchId, tx_in_account, ...)

ConstanceBeguier committed:
Update Swap Bundle figure

Replace split notes by reference notes
Replace timeLimit by expiryHeight

vivek-arte committed:
[Swaps] Updates to the Asset Swaps fees rationale and specification (#63)

This PR makes clearer the Fees section of ZIP 228, and also adds a
rationale section for the fees.
It also changes the specification of the computation of the spend
authorization signature.
It also adds Alexey Koren and Constance Beguier to the ZIP Owners.

vivek-arte committed:
Refactor of ZIP 228 (#128)

This PR goes through ZIP 228, and makes changes to reconcile the ZIP
with the updates made in ZIP 226, ZIP 227, ZIP 230 and ZIP 246, for the
NU7 protocol upgrade.

vivek-arte committed:
Merge branch 'main' into asset_swaps

vivek-arte committed:
Merge branch 'main' into asset_swaps

vivek-arte committed:
Merge branch 'main' into asset_swaps

vivek-arte committed:
Merge branch 'main' into asset_swaps

daira commented on static/assets/images/zip-0228-swap-bundle.png:
Text is a little small.

daira commented on zips/zip-0228.rst:
This will need thorough security analysis.

daira commented on zips/zip-0228.rst:
I don't think this prevents an adversarial matcher from adding actions to an action group. (The binding signature is over the overall txid, so it wouldn't ensure nonmalleability of an action group even if $\mathsf{bsk}$ were not revealed to the matcher.)

daira commented on zips/zip-0228.rst:
```suggestion
The $\mathsf{bsk}$ is only sent within the Swap Order. The Swap Order is assumed to be communicated over a secure channel off-chain between the parties. 
```

daira reviewed (approved):
ACK with @arya2 and @SamHSmith.
